python/pyspark/worker.py

```diff
@@ -111,9 +114,16 @@ def wrapped(key_series, value_series):
                 "Number of columns of the returned pandas.DataFrame "
                 "doesn't match specified schema. "
                 "Expected: {} Actual: {}".format(len(return_type), len(result.columns)))
-        arrow_return_types = (to_arrow_type(field.dataType) for field in return_type)
-        return [(result[result.columns[i]], arrow_type)
-                for i, arrow_type in enumerate(arrow_return_types)]
+        try:
+            # Assign result columns by schema name
+            return [(result[field.name], to_arrow_type(field.dataType)) for field in return_type]
+        except KeyError:
+            if all(not isinstance(name, basestring) for name in result.columns):
```

Comment:
Can we just do `isinstance(name, str)` here to deal with python2/3?

Reply:
I believe he's trying to deal with `unicode` case too just in python 2. `isinstance(name, basestring)` should be safer.

Reply:
Yeah, we still need to check for the possibility that python 2 uses unicode.